File: dynamic_programming/Min Cost Climbing Stairs.java
// https://leetcode.com/problems/min-cost-climbing-stairs/
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        Map<Integer, Integer> map = new HashMap<>();
        return minCostClimbingStairs(map, cost, cost.length, 0);
    }

    public int minCostClimbingStairs(Map<Integer, Integer> map, int[] cost, int num, int result) {
        if (num == 0 || num == 1) {
            return result;
        }
        if (map.containsKey(num)) {
            return map.get(num);
        }
        int result1 = minCostClimbingStairs(map, cost, num - 1, cost[num - 1]);
        int result2 = minCostClimbingStairs(map, cost, num - 2, cost[num - 2]);

        result = result + Math.min(result1, result2);
        map.put(num, result);
        return result;
    }
}

File: dynamic_programming/Min Cost Climbing Stairs.py
# https://leetcode.com/problems/min-cost-climbing-stairs/description
from functools import cache
from typing import List


class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        @cache
        def dfs(i: int):
            if i >= len(cost):
                return 0
            return cost[i] + min(dfs(i + 1), dfs(i + 2))

        return min(
            dfs(0), dfs(1)
        )  # You can either start from the step with index 0, or the step with index 1.


s = Solution()
print(s.minCostClimbingStairs(cost=[10, 15, 20]))
print(s.minCostClimbingStairs(cost=[1, 100, 1, 1, 1, 100, 1, 1, 100, 1]))
